### ISSUE TITLE
EndOfStreamException

### ISSUE THREAD

**Issue by Fluoxetinum** [DarkWanderer/ClickHouse.Client#184]
Hello,

I have a suspicion that this condition:
* https://github.com/DarkWanderer/ClickHouse.Client/blob/master/ClickHouse.Client/Formats/ExtendedBinaryReader.cs#L33 

is not always false.
In some cases, _bytesRead_ is less than _count_. 

I prepared demo app to demonstrate this - https://github.com/Fluoxetinum/ClickHouseClientEndOfStream. I created a clickhouse instance on Yandex cloud. You can find table schema in file 'scripts/table.sql'. Demo data is described in file 'scripts/query.sql'.

The app connects to the remote clickhouse instance (this bug does not occur when you connect to a local clickhouse instance) and makes select requests in infinite loop. Several seconds later, an exception is thrown. For example:

_Unhandled exception. System.IO.EndOfStreamException: Expected to read 8 bytes, got 6
   at ClickHouse.Client.Formats.ExtendedBinaryReader.Read(Byte[] buffer, Int32 index, Int32 count)
   at ClickHouse.Client.Types.UuidType.Read(ExtendedBinaryReader reader)
   at ClickHouse.Client.Types.ArrayType.Read(ExtendedBinaryReader reader)
   at ClickHouse.Client.ADO.Readers.ClickHouseDataReader.Read()
   at System.Data.Common.DbDataReader.ReadAsync(CancellationToken cancellationToken)_
![image](https://user-images.githubusercontent.com/47554699/186736463-ca5a6cc4-4147-4dd8-ba61-01294bb68411.png)

If remaining bytes are read with code like:
```
        public override int Read(byte[] buffer, int index, int count)
        {
            var countRemaining = count;
            var currentIndex = index;
            var bytesRead = base.Read(buffer, index, count);
            if (bytesRead == 0)
            {
                throw new EndOfStreamException($"Expected to read {countRemaining} bytes, got {bytesRead}");
            }
            countRemaining -= bytesRead;
            currentIndex += bytesRead;
            while (countRemaining > 0)
            {
                bytesRead = base.Read(buffer, currentIndex, countRemaining);
                if (bytesRead == 0)
                {
                    throw new EndOfStreamException($"Expected to read {countRemaining} bytes, got {bytesRead}");
                }
                countRemaining -= bytesRead;
                currentIndex += bytesRead;
            }

            return count;
        }
```
the exception will disappear.

Documentation for BufferedStream (https://docs.microsoft.com/en-us/dotnet/api/system.io.bufferedstream.read?view=net-6.0#system-io-bufferedstream-read(system-byte()-system-int32-system-int32) also refers to this aspect:

> The total number of bytes read into array. This can be less than the number of bytes requested if that many bytes are not currently available, or 0 if the end of the stream has been reached before any data can be read.

**Comment by DarkWanderer** [comment]
Thank you very much for detailed bug report and especially providing a test stand to reproduce, it is immencely helpful :+1:

You are correct that a `Stream` can return 0 only if there is no more data, however, it seems that `BinaryReader.Read` does not have such guarantee, so I opted for using `PeekChar()` instead. Version [5.0.4](https://github.com/DarkWanderer/ClickHouse.Client/releases/tag/5.0.4) is out now, please check if it fixes your issue (it does in my small-scale testing

**Comment by DarkWanderer** [comment]
Confirmed fixed in 5.0.4
